### ISSUE TITLE
Can this be used to predict disparity maps given stereo images?

### ISSUE THREAD

**Issue by raphaelchang** [nianticlabs/monodepth2#92]
Hello, thank you for this work and well documented code. Our project uses a stereo pair and we are trying to find a network that can perform stereo matching to compute disparity maps (for metric depth). We do not have ground truth disparity maps so monodepth2 seemed appealing. I have a few questions regarding the suitability of monodepth2 for our project.

1. I know it is designed to infer on a single image and predict depth. Since the network learns the disparity map, do you know if it is possible to make some modifications so that it can take two images and return the disparity map?
2. I see that the stereo baseline is hardcoded to 0.1 in the dataloader. Does this mean it can be trained on stereo data with varying baselines? How does the disparity output know which baseline the disparity is for?
3. I understand that the output is inverse depth which is proportional to disparity. What is the scale of this if it is trained on stereo data? If we know the baseline at training and/or runtime, is there a way to get metric disparity/depth? What happens if your training baseline/focal length is different from your runtime ones?

**Comment by Sibozhu** [comment]
This question seems so valuable, I also want to know that!

**Comment by daniyar-niantic** [comment]
Hi @raphaelchang 

1. Monodepth is designed to take single image and produce depth map. If you are interested in stereo matching from a stereo pair, please consider using other methods such as: 
https://github.com/feihuzhang/GANet
https://github.com/JiaRenChang/PSMNet

2. The baseline is hardcoded for KITTI, it is there mostly for training stability, see https://github.com/nianticlabs/monodepth2/issues/30

The network never knows anything about the baseline because it only sees one image. Your question about varying disparity is more relevant for stereo matching methods.

3. Please checkout these:
https://github.com/nianticlabs/monodepth2/issues/61
https://github.com/nianticlabs/monodepth2/issues/66

**Comment by raphaelchang** [comment]
Hi @daniyar-niantic 

Thank you for your quick response. I have a few follow up questions.

1. If we wanted monodepth2 to output metric inverse depth, is that possible? If we train on stereo pairs and provide a known baseline at training, does this allow monodepth2 to predict metric depth?
2. My question about varying baselines was intended to apply to the training on stereo pairs. Does all the training data need to come from the same baseline and focal length? Or if we dynamically set stereo_T can it be varying (assuming the weights converge)? Also see question 3.
3. If I understand from #30 and evaluate_depth.py, the baseline is hardcoded to 0.1 and then scaled to the correct baseline at evaluation to get metric depth. Does this mean the model has to be trained on a consistent baseline, in order to be scaled back to the correct scale at runtime? Or if we dynamically set stereo_T at runtime, can we get metric depth without manually scaling at runtime?

**Comment by daniyar-niantic** [comment]
Yes, everything you outlined seems possible. But it has to be verified experimentally as you are going into uncharted territory.

I think that varying baseline is not a problem, as long as intrinsics don't change significantly.
If the intrinsics change, you should check http://openaccess.thecvf.com/content_CVPR_2019/papers/Facil_CAM-Convs_Camera-Aware_Multi-Scale_Convolutions_for_Single-View_Depth_CVPR_2019_paper.pdf
and
https://arxiv.org/abs/1904.04998

The conversion that transforms the network's sigmoid output to depth can be in any units. If you know the baseline, you can work out what are the depth units in meters. So, the network output would be metric.

4. Is your question about generalisability of any machine learning model? So, this question can be asked of all neural networks?
